model/routing/ResourceUrlBuilder.php

```diff
@@ -15,81 +15,87 @@
 
-use oat\generis\model\OntologyAwareTrait;
-use oat\oatbox\service\ConfigurableService;
+use LogicException;
+use oat\tao\model\menu\Tree;
+use InvalidArgumentException;
+use core_kernel_classes_Class;
+use oat\tao\model\menu\Section;
+use core_kernel_classes_Resource;
 use oat\tao\model\menu\MenuService;
 use oat\tao\model\menu\Perspective;
-use oat\tao\model\menu\Section;
+use oat\oatbox\service\ConfigurableService;
 
-/**
- * @author Gyula Szucs <gyula@taotesting.com>
- */
 class ResourceUrlBuilder extends ConfigurableService
 {
-    use OntologyAwareTrait;
-
-    const SERVICE_ID = 'taoBackOffice/resourceUrlBuilder';
+    public const SERVICE_ID = 'taoBackOffice/resourceUrlBuilder';
 
     /**
      * Builds a full URL for a resource
      *
-     * @param \core_kernel_classes_Resource $resource
      * @return string
      */
-    public function buildUrl(\core_kernel_classes_Resource $resource)
+    public function buildUrl(core_kernel_classes_Resource $resource)
     {
-        if (!$resource->exists()) {
-            throw new \InvalidArgumentException('The requested resource does not exist or has been deleted');
-        }
+        $resourceClass = $resource->getClass($resource);
 
-        /** @var Perspective $perspective */
-        /** @var Section $section */
+        if (!$resource->exists() && !$resourceClass->exists()) {
+            throw new InvalidArgumentException('The requested resource does not exist or has been deleted');
+        }
 
-        $resourceClass = $resource->isClass()
-            ? $this->getClass($resource)
-            : array_values($resource->getTypes())[0];
+        if (!$resource->isClass()) {
+            $resourceClass = array_values($resource->getTypes())[0];
+        }
 
+        /** @var Perspective $perspective */
         foreach (MenuService::getAllPerspectives() as $perspective) {
+            /** @var Section $section */
             foreach ($perspective->getChildren() as $section) {
                 if ($this->isSectionApplicable($resourceClass, $section)) {
                     return $this->getBackofficeUrl($perspective, $section, $resource);
                 }
             }
         }
 
-        throw new \LogicException('No url could be built for "' . $resource->getUri() . '"');
+        throw new LogicException(
+            sprintf(
+                'No url could be built for "%s"',
+                $resource->getUri()
+            )
+        );
     }
 
     /**
      * Generates the actual URL based on perspective and section
-     * @param Perspective $perspective
-     * @param Section $section
-     * @param \core_kernel_classes_Resource $resource
+     *
+     * @return string
      */
-    private function getBackofficeUrl(Perspective $perspective, Section $section, \core_kernel_classes_Resource $resource)
-    {
+    private function getBackofficeUrl(
+        Perspective $perspective,
+        Section $section,
+        core_kernel_classes_Resource $resource
+    ) {
```

Comment:
suggestion: I think you can directly put a return value here instead of the PHPDoc annotation (as in the next method).
```suggestion
    ): string {
```

Reply:
I don't want to do this, because I'm not sure about other of the code inside this method - all of them doesn't have return types, so I cannot be sure, that this will return a string. 🙂

Reply:
`_url` just calls `tao_helpers_Uri::url` and returns its value, and the code for that method always return a string (it returns the concatenation of some strings). BTW I get your point, so consider this just as a suggestion :) 

Marking this thread as solved.